Add netmon support

rajkan01 opened the pull request:
- Add sntp client implementation to connect to SNTP server to get real-time in seconds since 1st Jan 1970 and also regularly sync SNTP timestamp on 24hours.
- Connect to the SNTP server on net connect to initialize the time base.
- Regularly sync SNTP timestamp when connected (configurable interval)
- Monitor connection state, when the connection goes down cancel the sntp time sync and await for the network.
- move the network DHCP wait implementation to the network directory.
- Add the get the network state API which helps to find the network state, like up or down by other threads.
- Update CMake.
- Update Kconfig and overlay network conf
- Add netmon stats support using zephyr stats API for tracking  DHCP up/down events, SNTP requests, and SNTP reads counts.
- netmon to update stats on DHCP up/down events and sntp resync time events. SNTP client to update stats on all sntp time requests.
- Add the network status atomic variable, which helps to identify the network condition, and it can be read and modified by threads in an uninterruptible manner to find the network condition.

Commit:
Sample: add SNTP client support

Add sntp client implementation to connect to SNTP server to get
real-time in seconds since 1st Jan 1970 and also regularly sync SNTP
timestamp on 24hours.

Signed-off-by: Rajkumar Kanagaraj <rajkumar.kanagaraj@linaro.org>

Commit:
sample: Network monitor management thread

- Connect to the SNTP server on net connect to initialize the time base.
- Regularly sync SNTP timestamp when connected (configurable interval)
- Monitor connection state, when the connection goes down cancel the
  sntp time sync and await for the network.

Signed-off-by: Rajkumar Kanagaraj <rajkumar.kanagaraj@linaro.org>

Commit:
Sample: move DHCP wait

- move the network DHCP wait implementation to the network directory.
- Add the get the network state API which helps to find the network
  state, like up or down by other threads.
- Update CMake.

Signed-off-by: Rajkumar Kanagaraj <rajkumar.kanagaraj@linaro.org>

Commit:
sample: Update Kconfig and overlay network conf

Signed-off-by: Rajkumar Kanagaraj <rajkumar.kanagaraj@linaro.org>

Commit:
sample: netmon: Add netmon stats support

Add netmon stats support using zephyr stats API for tracking  DHCP
up/down events, SNTP requests, and SNTP reads counts.

Signed-off-by: Rajkumar Kanagaraj <rajkumar.kanagaraj@linaro.org>

Commit:
sample: network: update netmon and sntp client with stats calls

netmon to update stats on DHCP up/down events and sntp resync time
events. SNTP client to update stats on all sntp time requests.

Signed-off-by: Rajkumar Kanagaraj <rajkumar.kanagaraj@linaro.org>

d3zd3z commented on src/network/sntp_client.c:
```diff
@@ -0,0 +1,105 @@
+}
+
+void sntp_client_init()
+{
```
As I understand, this API maintains it's own time base. Would we be better off working with the existing and posix time functions, which already have a mechanism for dealing with a base.

Regardless of this, are there any concerns about time skips that come about because the system clock runs at a different speed than the real world time? It is possible to get either a discontinuity, or for time to appear to go backward.

Also, how does `k_uptime_get` respond to systems that enter low power mode?

d3zd3z commented on src/network/netmon.c:
```diff
@@ -0,0 +1,70 @@
+ */
+static void netmon_time_sync(struct k_work *work)
+{
+
```
No blank line at beginning of function.

d3zd3z commented on src/network/netmon.c:
```diff
@@ -0,0 +1,70 @@
+
+	if (!is_dhcp_up()) {
+		LOG_ERR("Network down");
+		goto end;
```
I think this works, but let me make sure things are clear. We re-schedule the timer for 24 hours when the network is down. But the monitoring thread will also notice the network is down, and cancel this timer, and schedule a new one for a shorter interval?

d3zd3z commented on src/network/netmon.c:
```diff
@@ -0,0 +1,70 @@
+		LOG_ERR("Network down");
+		goto end;
+	}
+	sntp_resync_time();
```
Should we handle the case that the resync fails, retry earlier?

d3zd3z commented on src/network/netmon.c:
```diff
@@ -0,0 +1,70 @@
+
+	/* Reschedule the worker therad to resync the time base with SNTP server. */
+end:
+	k_work_reschedule(&time_sync, K_HOURS(24));
```
Any concerns that this will drift as it will generally be slightly more than 24 hours?

d3zd3z commented on src/network/dhcpwait.c:
```diff
@@ -37,6 +37,11 @@ void await_dhcp(void)
 
+bool is_dhcp_up()
+{
+	return dhcp_running;
```
I didn't really think about this before, and probably doesn't matter on the coretex CPUs we're on, but technically, this should be an atomic variable, or there could be risks of incoherence between the different threads reading and writing it.

microbuilder replied:
There is support for atomic variables in Zephyr, at least on Arm from memory 👍

Edit: https://docs.zephyrproject.org/latest/kernel/services/other/atomic.html

microbuilder replied:
@d3zd3z What do you think is the best synchronization mechanism in this use case in Zephyr to indicate a state change across threads? Zephyr has several options for this with various pros and cons. I did look at the docs, but curious if you have a strong opinion here?

d3zd3z commented on src/network/sntp_client.c:
```diff
@@ -0,0 +1,105 @@
+	// 	LOG_ERR("Failed to set system time, %d", ret);
+	// }
+
+    /* Increment the sntp time req stats counter */
```
This seems to be indented by 4 spaces, not tabs.

Please make sure your editor is configured to use tabs for indentation, and to display tabs as 8 spaces, in order to get the right formatting for Zephyr's coding style.

microbuilder replied:
Just use clang-format and the config file in the root of Zephyr. If you're on VSCode you can install the clang-format plugin, and for your local project point formatting to that plugin, and reformat the whole file for Zephyr's style in one keyboard shortcut. CI uses the same file to verify formatting in PRs.

d3zd3z commented on src/network/sntp_client.c:
```diff
@@ -0,0 +1,105 @@
+	LOG_INF("Data and time: %04u-%02u-%02u %02u:%02u:%02u",
+		info.tm_year + 1900, info.tm_mon + 1, info.tm_mday,
+		info.tm_hour, info.tm_min, info.tm_sec);
+
```
Extra blank line at end of function.

d3zd3z reviewed (changes requested):
Looks good, just a few minor things.

Commit:
sample: Add network status atomic variable support

Add the network status atomic variable, which helps to identify the
network condition, and it can be read and modified by threads in an
uninterruptible manner to find the network condition.

Signed-off-by: Rajkumar Kanagaraj <rajkumar.kanagaraj@linaro.org>

microbuilder commented on include/netmon_stats.h:
```diff
@@ -0,0 +1,80 @@
+	(void)(net_sts);
+	(void)(stats);
+}
+#endif /* CONFIG_NETMON_STATS */
```
This might be cleaner to only have one instance of the functions, and have the bodies in an `#ifdef` check, rather than two copies of the function signature?

microbuilder commented on src/network/netmon_mgmt.c:
```diff
@@ -0,0 +1,69 @@
+ */
+static void netmon_mgmt_time_sync(struct k_work *work)
+{
+
```
Extra line ... if you haven't, please run these through `clang-format` with Zephyr's config file for that, although maybe it just doesn't catch the leading space issues.

microbuilder commented on src/network/netmon_mgmt.c:
```diff
@@ -0,0 +1,69 @@
+	netmon_sntp_time_sync_call_stats(get_netmon_stats_inst());
+
+	/* Reschedule the worker therad to resync the time base with SNTP server. */
+	k_work_reschedule(&time_sync, K_HOURS(24));
```
Maybe make this a KConfig variable, with default 24?

microbuilder commented on src/network/netmon_mgmt.c:
```diff
@@ -0,0 +1,69 @@
+			await_dhcp();
+			netmon_dhcp_stats(true, get_netmon_stats_inst());
+			// TODO TLS re-intialization
+			k_work_reschedule(&time_sync, K_SECONDS(1));
```
Better to be consistent with the comment format ... Zephyr coding standards dictate always using `/* .... */` from memory.

microbuilder reviewed (changes requested):
Just some cosmetic things, but good to clean them up now. Thanks for the changes so far, though.

Commit:
sample: netmon: Add Kconfig variable for resync time

Signed-off-by: Rajkumar Kanagaraj <rajkumar.kanagaraj@linaro.org>

Commit:
sample: netmon_stats: make one function instance

Rather than have a duplicate dummy implementation for netmon stats APIs
in case of not enabling the netmon stats to support, update a single
function instance with ifdef to control the flow accordingly.

Signed-off-by: Rajkumar Kanagaraj <rajkumar.kanagaraj@linaro.org>

microbuilder commented on Kconfig:
```diff
@@ -68,4 +68,31 @@ config NV_COUNTER_TRACKER_THRESHOLD_LIMIT
+	help
+	  Size of the stack to be used by the netmon worker thread.
+
+config SNTP_INIT_SERVER
```
`INIT` doesn"t really make sense to me here. Maybe just:
```suggestion
config SNTP_SERVER
```

microbuilder replied:
Or `SNTP_SERVER_NAME`

microbuilder commented on include/netmon_stats.h:
```diff
@@ -0,0 +1,72 @@
+STATS_NAME(netmon, sntp_time_req_count)
+STATS_NAME_END(netmon);
+#else
+struct stats_netmon {
+};
```
Where is `struct stats_netmon` defined when `CONFIG_NETMON_STATS=y`? Does the stats subsys create it for us?

rajkan01 replied:
yes it is created when CONFIG_NETMON_STATS=y

Commit:
sample: netmon: Rename SNTP server Kconfig

Signed-off-by: Rajkumar Kanagaraj <rajkumar.kanagaraj@linaro.org>
